### ISSUE TITLE
CSS Color; light-dark() should be applicable on more properties.

### ISSUE THREAD

**Issue by MattWilcox** [w3c/csswg-drafts#10577]
Please forgive me if this one's a bit vague as I'm adding an issue here at Chris Lilley's request, and I'm not familiar with the process too much.

There are times when I need to swap the background-image as well as text colour, and it's 100% dependant on the same logic as light-dark() value. But, you can't do this:

```css
.thing {
  background-image: light-dark(
    url('/assets/light.svg'),
    url('/assets/dark.svg')
  );
}
```

I'd like to be able to do that instead of having to resort to media queries to do what in the end is the same thing, for the same reason, at the same time - but not working on a colour property. In my mind, `light-dark()` is a toggle to map key-value pairs based on a scheme - but does not need to be (and should not be) tied to outputting "colour" as a return value (or however this is currently implemented/specced).

In this specific case, I have a graphical decoration added to some link text, and need to change the decoration's colour to match the text colour. Because SVG images loaded in via CSS don't work the same as if they're embeded in HTML they will not inherit to text colour, and I have to specify an entirely different asset to load to match the text colour.

https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/light-dark

https://mstdn.social/@mattwilcox/112792513614070547

**Comment by tabatkins** [comment]
Yeah, in the discussions for `light-dark()` we discussed the need for a more general function that can return any value (provsionally called `schemed-value()`). It was put to the side pending more use-cases, because it requires giving it `var()`-like semantics, which isn't as good as a function with a proper type (like `light-dark()`, which is a `<color>`).

**Comment by svgeesus** [comment]
The use case (pointing to separate light and dark media versions) is certainly a strong one. Especially for SVG images, which for security reasons are forbidden to link to external stylesheets, fonts etc or to inherit the surrounding styling context, when used in CSS or in the HTML `<img>` element.

**Comment by MattWilcox** [comment]
I'm not read in on the thinking behind all of this, so take this with a grain of salt - but I will say from the perspective of someone using the language without having read the spec, I don't find `light-dark()` being cast to a type useful at all. Nothing about the language used for the function name indicates it would or should return a colour type. From my perspective, its just a thing that should select one value or the other based on the match. Now, if it had been called `color-light-dark()` then yeah, I'd expect that limitation.

**Comment by SebastianZ** [comment]
What if we extended `light-dark()` for this use case, which seems absolutely reasonable, and continue discussing a more general solution in #7561?
I.e. I propose a variant of `light-dark()` that takes two `<image>` values and returns an `<image>` value.

Sebastian

**Comment by tabatkins** [comment]
We can't extend it. It being a `<color>` means it can be grammar-checked in 'color', in gradients, etc. If it can be multiple types, then we can't grammar-check it and have to treat it like `var()` - that's what I was referring to.

If you meant "add a second function that's just like light-dark() but for images", at that point we might as well do the full thing. It's not meaningfully *harder*.

**Comment by SebastianZ** [comment]
> If you meant "add a second function that's just like light-dark() but for images", ...

That was my idea, yes. A function with the same name but for images instead of a color.

> ... at that point we might as well do the full thing. It's not meaningfully _harder_.

Meaning giving it `var()`-like semantics, as you wrote earlier, I assume. If that's the case, then let's aim for the full thing. I just thought restricting it to just `<image>` values would make it easier to spec. and implement.

> ... which isn't as good as a function with a proper type (like `light-dark()`, which is a `<color>`).

Maybe not from a Typed OM point of view, though from pure CSS view it does, IMO.

Sebastian

**Comment by svgeesus** [comment]
> If you meant "add a second function that's just like light-dark() but for images", at that point we might as well do the full thing. It's not meaningfully _harder_.

Okay so lets do that, then.

> I propose to go for `light-dark(<foo>, <bar>)`, then extend to `schemed-color` or `schemed-value` in the future, if needed?

> I'd propose to do this for `<color>` (or `<color>` and `<image>`) for now, rather than everything.

_Originally posted by @emilio in https://github.com/w3c/csswg-drafts/issues/7561#issuecomment-1644625985_

**Comment by svgeesus** [comment]
@bramus also [blogged about why we went for `light-dark()`](https://www.bram.us/2023/10/09/the-future-of-css-easy-light-dark-mode-color-switching-with-light-dark/#schemed-value) but seemed to argue that a more general function was gated behind custom color-scheme support.

I don't see why that has to be the case.

**Comment by romainmenke** [comment]
Is there sufficient overlap with https://github.com/w3c/csswg-drafts/issues/10064 to consider something like :

`color-scheme(dark)` as a condition for `<<if-cond>>` ?

So that you can do : `if(color-scheme(dark): blue, else: cyan)`

In other words do we need a dedicated function for color schemed values if we are already considering other syntax for more generic inline conditions?

**Comment by tabatkins** [comment]
That is a very good point! It's a little longer, but it does neatly solve the parsing issues we'd have to address in a generic schemed-value() (which don't apply to light-dark(), since we know the value is a color).

**Comment by svgeesus** [comment]
> So that you can do : `if(color-scheme(dark): blue, else: cyan)`

So, what would that look like for two image  urls such as `foo-dark.svg` and `foo-light.svg` ?

**Comment by tabatkins** [comment]
`if(color-scheme(dark): url(foo-dark.svg), else: url(foo-light-svg))`

**Comment by svgeesus** [comment]
Okay, re-tagging to https://github.com/w3c/csswg-drafts/labels/css-values-5 because `if()` seems to be the way forward

**Comment by svgeesus** [comment]
@MattWilcox thanks for raising this, and I hope that the current direction meets your needs even though it doesn't use your proposed solution. I think it is a good way forward, and a solution is certainly needed.

**Comment by nmn** [comment]
Hi! I wanted to chime in about expanding `light-dark()` itself to support any type, as mentioned in this interop issue. (I've been told to discuss it here instead)

https://github.com/web-platform-tests/interop/issues/841#issuecomment-2401610759

---

>  It was put to the side pending more use-cases, because it requires giving it var()-like semantics, which isn't as good as a function with a proper type (like light-dark(), which is a <color>).

Why is it not possible to give `light-dark()` semantics more similar to  `calc` instead? Instead of `light-dark()` being `any`-typed, it should be the same type as the arguments within.

- `light-dark(<color>, <color>)` has the type of `<color>`
- `light-dark(<url>, <url>)` has the type `<url>`
- ... and so on

If this is possible it would solve the regressions caused by opening up `light-dark()` to non color values.

The use-case I want more than anything else is:

```css
color-scheme: light-dark(dark, light);
```

I.e., I want to be able to "flip" the current color scheme within a particular HTML element.

**Comment by romainmenke** [comment]
> The use-case I want more than anything else is:
> 
> ```css
> color-scheme: light-dark(dark, light);
> ```

🤔 I think this would be cyclic and should be invalid.

Same for:
```css
color-scheme: if(color-scheme(dark): dark, else: light);
```

**Comment by nmn** [comment]
@romainmenke Oh! I see that `color-scheme` applies on the same element just like CSS variables. Thanks for the correction. I'll use CSS custom properties to work around this issue.

The point of using it for other value types stands anyway.

**Comment by sarajw** [comment]
> `if(color-scheme(dark): url(foo-dark.svg), else: url(foo-light-svg))`

Ooooh. May I double check whether this would be querying the current state of what the browser is showing, and not just what we've given to `color-scheme:`?

What I mean is:
A computer operating system is in dark mode, the user opens a web page with `color-scheme: light dark;` declared in its CSS (or the equivalent in a meta tag) and displays the dark version of the page. Would `if(color-scheme(dark) ... )` come back true ?

**Comment by tabatkins** [comment]
Yeah, the 'color-scheme' property doesn't *set* any color scheme, it just tells the browser which schemes are acceptable to use on the element. The condition would be querying the actual scheme the element is using.

**Comment by o-t-w** [comment]
> Because SVG images loaded in via CSS don't work the same as if they're embeded in HTML they will not inherit to text colour, and I have to specify an entirely different asset to load to match the text colour.

A better solution would be to use `light-dark()` [within the SVG](https://fullystacked.net/dark-mode-svg/) itself.

**Comment by MattWilcox** [comment]
It's not "better" if I now have to manage duplicate code in two different locations. Or multiply that up for as many SVG assets as are needing to be effected.

**Comment by emilio** [comment]
FWIW while on SVG this is indeed doable (via prefers-color-scheme inside, or so), this seems a fair use-case, and the `light-dark()` function (even if it's effectively syntax sugar for `if(color-scheme(...), ...)`) might still be useful? (I just came across a use case for this in the Firefox front-end for example).

**Comment by tabatkins** [comment]
Yeah, `light-dark()` for colors, specifically, is such a common use-case that I think it justifies its existence as sugar. I'm happy to leave the "light-dark() but for any value type" function as just being handled by `if()`, tho.

**Comment by tabatkins** [comment]
For clarity, the Agenda+ proposal is to add `color-scheme()` to `<if-test>`, which tests the used color scheme on the element.

**Comment by emilio** [comment]
Well, I meant `light-dark()` for images is probably common as well, but if you don't think that's needed. Also, curious, does `color-scheme()` use the color-scheme property of the parent (like style queries?) or the element's?

**Comment by bramus** [comment]
> Also, curious, does `color-scheme()` use the color-scheme property of the parent (like style queries?) or the element's?

That is answered [here](https://github.com/w3c/csswg-drafts/issues/10577#issuecomment-2448061452): _“The condition would be querying the actual scheme the element is using.”_

On a side-note: You can hack your way into a `light-dark()` that works for _any value_ by combining custom functions and `if()`. See https://brm.us/at-function-and-if for details.

**Comment by LeaVerou** [comment]
Besides images, other use cases involve numbers or keywords. For example, you may want bolder text in dark mode, or you may want to define a color with one color component varying based on color scheme.

I wonder if we could introduce a third argument to `light-dark()` that is modeled after `attr()`'s `<attr-type>` and defaults to color.

**Comment by tabatkins** [comment]
For the more generic case, `if()` handles things. `font-weight: if(color-scheme(light): normal; else: bold);`. The full generic version of the function is different from the "just color" or even "just color or image" that is currently specified.

**Comment by tabatkins** [comment]
(Reminder once again, because it's now buried in the "Expand N comments" section: this issue is Agenda+ to ask for approval to add the `color-scheme()` test if `if()`.)

**Comment by tabatkins** [comment]
Actually, this might be good for an async resolution, since it seems uncontroversial and nobody seems to object in this thread. @astearns, sound reasonable?

* Proposed async resolution: add `color-scheme()` test to `if()`, which tests the used color scheme on the element

**Comment by LeaVerou** [comment]
@tabatkins Presumably the resolution is not `if()` specific and applies to container queries as well, right?

Looking at the [spec](https://drafts.csswg.org/css-values-5/#if-notation), it's unfortunate that these need to be updated separately, because experience shows they _will_ get out of sync. Can we define a shared token that they both reference? Then they can only add any additional stuff in their own definitions. (Let me know if I should open a new editorial issue on that)

**Comment by SebastianZ** [comment]
Just for the record, you can already achieve this via `if(media(prefers-color-scheme: dark): …)`.

Sebastian

**Comment by emilio** [comment]
That's not equivalent tho, because that doesn't account for the `color-scheme` property.

**Comment by astearns** [comment]
The CSSWG will automatically accept this resolution one week from now if no objections are raised here. Anyone can add an emoji to this comment to express support. If you do not support this resolution, please add a new comment.

Proposed Resolution: add `color-scheme()` test to `if()` and media queries, which tests the used color scheme on the element

**Comment by romainmenke** [comment]
@astearns @tabatkins The proposed resolution should probably be:

> Proposed Resolution: add `color-scheme()` test to `if()` and **container** queries, which tests the used color scheme on the element

Although arguably adding it to `container` queries requires more thought. Is there an actual use case for querying the used color scheme of a container as this can be different from the element itself? Don't feel very strongly about this, having parallel between container queries and `if()` does have benefits.

**Comment by fantasai** [comment]
I'm in favor of the resolution, but I would prefer that new features (which are by definition not trivial decisions) go through the telecon agenda, even if it's very briefly, just to make sure they get attention.

(Relatedly, and I've said this before, I think all async resolution requests should be listed on the main agenda in an appendix.)

**Comment by tabatkins** [comment]
> Although arguably adding it to container queries requires more thought.

Right, which is why I'm not proposing to add it as a CQ right now. ^_^

**Comment by astearns** [comment]
@LeaVerou do you have any concerns about limiting the addition to `if()` for now?

**Comment by LeaVerou** [comment]
> [@LeaVerou](https://github.com/LeaVerou) do you have any concerns about limiting the addition to `if()` for now?

No, that makes sense.

**Comment by mirisuzanne** [comment]
I do think it would be useful to have as a container query. There are cases where you need to account for both internal contrast in an element, and also how the element contrasts with it's broader context. 

I also agree with @LeaVerou that there's a lot of overlap in the conditions we'll want across `@container` and `if()` – so we should clarify how they relate, and how to provide features that work across both. They are not equivalent in what target element they query – `if()` is a self-query, while `@container` is an ancestor-query – but I think any shared syntax should always have the same shape to get the same info from the target. I think that's true with `style()` and should continue to be the pattern for other functions that work in both places.

Which I think should include this one. I don't feel strongly that they _have to_ happen at the same time, but I do think we should answer that question about how to keep these two conditional features in sync sooner rather than later. And I might argue that the default should be for features to work in both, and we should need good evidence to support only one or the other.

**Comment by tabatkins** [comment]
> And I might argue that the default should be for features to work in both, and we should need good evidence to support only one or the other.

I think that's totally reasonable.

**Comment by astearns** [comment]
Let’s take this to a regular meeting and talk about how to keep things in sync, and what the first step here should include.

**Comment by johannesodland** [comment]
> [@astearns](https://github.com/astearns) [@tabatkins](https://github.com/tabatkins) The proposed resolution should probably be:
> 
> > Proposed Resolution: add `color-scheme()` test to `if()` and **container** queries, which tests the used color scheme on the element
> 
> Although arguably adding it to `container` queries requires more thought. Is there an actual use case for querying the used color scheme of a container as this can be different from the element itself? Don't feel very strongly about this, having parallel between container queries and `if()` does have benefits.

Having it as a **media** query would make it possible to provide matching sources for `<picture>` and `<video>` sources. It would probably have to be based on the used color-scheme on the root element.

**Comment by romainmenke** [comment]
> Having it as a media query would make it possible to provide matching sources for `<picture>` and `<video>` sources. It would probably have to be based on the used color-scheme on the root element.

But being able to do `if()` checks or `container queries` in `<picture>` and `<video>` elements is the actual solution to that and many other issues, right? (wondering if this might be doable for lazy loaded images only, comparable to `sizes=auto`)

**Comment by johannesodland** [comment]
> > Having it as a media query would make it possible to provide matching sources for `<picture>` and `<video>` sources. It would probably have to be based on the used color-scheme on the root element.
> 
> But being able to do `if()` checks or `container queries` in `<picture>` and `<video>` elements is the actual solution to that and many other issues, right? (wondering if this might be doable for lazy loaded images only, comparable to `sizes=auto`)

That would not work on eagerly loaded images though, and they tend to be more important.

**Comment by theking2** [comment]
> Is there sufficient overlap with [#10064](https://github.com/w3c/csswg-drafts/issues/10064) to consider something like `color-scheme(dark)` as a condition for `<<if-cond>>` ?
> 
> So that you can do : `if(color-scheme(dark): blue, else: cyan)`
> 
> In other words do we need a dedicated function for color schemed values if we are already considering other syntax for more generic inline conditions?

why not `calc( color-scheme(dark)? blue: cyan )`

**Comment by stof** [comment]
> why not `calc( color-scheme(dark)? blue: cyan )`

because `calc` returns a dimension (i.e. a number with an optional unit), not an arbitrary type.
And `calc` does not support a ternary operator either. The `if()` function has already been settled previously (the topic of this issue is just the `color-scheme()` condition, not the whole existence of the `if` function).
